mesonbuild/modules/python.py

```diff
@@ -149,32 +162,101 @@ def extension_module_method(self, args: T.List['TYPE_var'], kwargs: 'TYPE_kwargs
 
             kwargs['install_dir'] = self._get_install_dir_impl(False, subdir)
 
+        target_suffix = self.suffix
+
+        limited_api_version = kwargs.pop('limited_api', '')
+        if not isinstance(limited_api_version, str):
+            raise InvalidArguments('"limited_api" argument must be a string.')
+
         new_deps = mesonlib.extract_as_list(kwargs, 'dependencies')
-        has_pydep = any(isinstance(dep, _PythonDependencyBase) for dep in new_deps)
-        if not has_pydep:
+        pydep = next((dep for dep in new_deps if isinstance(dep, _PythonDependencyBase)), None)
+        if pydep is None:
             pydep = self._dependency_method_impl({})
             if not pydep.found():
                 raise mesonlib.MesonException('Python dependency not found')
             new_deps.append(pydep)
             FeatureNew.single_use('python_installation.extension_module with implicit dependency on python',
                                   '0.63.0', self.subproject, 'use python_installation.dependency()',
                                   self.current_node)
+
+        if limited_api_version != '':
+
+            has_cython_files = any(source_file.endswith('.pyx') for source_file in args)
```

Comment:
`source_file` is not necessarily a `str`. It could be a `File`, a `CustomTarget`, etc. You cannot use `.endswith` directly on it.

Reply:
Thanks for finding this! 

I'm prototyping a fix to this and I am wondering if something like this would get the job done:

```python
def check_for_cython_files(args):
  for arg in args:
    if isinstance(arg, mesonlib.File) and arg.endswith('.pyx'):
      return True
    elif isinstance(arg, build.BuildTarget) and check_for_cython_files(arg.get_sources()):
      return True
  return False
```

Although maybe there exists functionality already for doing something like this that I'm not aware of?

I admit that the Cython support for this is considerably less well tested currently than raw C/C++. Cython's support for the Python Limited API is currently limited to the upcoming 3.0 release, which I wasn't able to find packaged anywhere that allowed easy testing. It might therefore be a good idea to just remove this functionality until this version of Cython is more widely available. I have no strong feelings either way, and will do whatever is amenable to maintainers.

Reply:
I think you want get_outputs, actually. A BuildTarget won't be supported at all here -- those are executable() or library() return values, and get_sources would be the inputs to the executable, rather than the inputs to py.extension_module. The concern here is CustomTarget instead, which might be used to produce .c or .pyx files at build time. You do still have to check for str.endswith as well.

It may of course not be worth making this check at all -- what's the outcome of a failure here?

Reply:
> It may of course not be worth making this check at all -- what's the outcome of a failure here?

The check exists to detect if we have any cython files as input so that we can then check to see if the version of Cython supports the limited API. If it doesn't we can then emit a warning. So the outcome of failure here (defined as failing to correctly detect the presence of cython files in the input to this extension module) would mean that a warning is not emitted when it should be.

Is the value of this warning worth the complexity of this task?

Reply:
Let's just say that I would in no way consider this any kind of blocker, especially if it's possible to get a false positive.